File: adapters/elixir/test/host_execution_observability_test.exs
defmodule PraugaFlexDoc.HostExecutionObservabilityTest do
  use ExUnit.Case, async: false

  alias PraugaFlexDoc.HostExecution
  alias PraugaFlexDoc.HostExecutionMetric
  alias PraugaFlexDoc.HostExecutionObservability
  alias PraugaFlexDoc.HostExecutionObservation

  defp collecting_executor(origins \\ ["https://api.example.test"]) do
    test = self()
    HostExecution.new!(origins, metric_sink: fn metric -> send(test, {:metric, metric}) end)
  end

  defp collected_metrics(acc \\ []) do
    receive do
      {:metric, metric} -> collected_metrics([metric | acc])
    after
      0 -> Enum.reverse(acc)
    end
  end

  defp find(metrics, name), do: Enum.find(metrics, &(&1.name == name))

  defp metric(name, kind, value, labels \\ %{}) do
    %HostExecutionMetric{name: name, kind: kind, value: value, labels: labels}
  end

  defp start_recorder(options \\ []) do
    {:ok, recorder} = HostExecutionObservation.start_link(options)
    recorder
  end

  # Casts are asynchronous, so read through a call to be sure every fold landed.
  defp settled_snapshot(recorder), do: HostExecutionObservation.snapshot(recorder)

  # The reason vocabulary is a cross-runtime contract, not an Elixir detail: a
  # collector written against any other FlexDoc host must read these labels
  # unchanged.
  test "reasons match the shared vocabulary" do
    assert HostExecutionObservability.reasons() == [
             "marker-missing",
             "execution-disabled",
             "admission-saturated",
             "destination-forbidden",
             "redirect-forbidden",
             "body-malformed",
             "body-too-large",
             "unsupported-media-type",
             "request-invalid",
             "auth-unsupported",
             "upstream-timeout",
             "upstream-unreachable",
             "upstream-error"
           ]

    assert HostExecutionObservability.reason?("upstream-timeout")
    refute HostExecutionObservability.reason?("slow")
  end

  test "an unmarked request is counted outside the lifecycle" do
    executor = collecting_executor()
    HostExecution.handle(executor, nil, %{})

    assert [metric] = collected_metrics()
    assert metric.name == "flexdoc_execute_unmarked_total"
    assert metric.labels["reason"] == "marker-missing"
  end

  test "a policy rejection carries its reason and is not an upstream error" do
    executor = collecting_executor()

    result =
      HostExecution.handle(executor, "1", %{
        "request" => %{"url" => "https://blocked.example.test/pets", "method" => "GET"}
      })

    assert result.status == 403

    metrics = collected_metrics()
    rejection = find(metrics, "flexdoc_execute_rejections_total")
    assert rejection.labels["reason"] == "destination-forbidden"
    assert rejection.labels["statusCode"] == "403"
    assert rejection.labels["source"] == "route"
    refute find(metrics, "flexdoc_execute_errors_total")
  end

  test "a malformed envelope is separable from a policy rejection" do
    executor = collecting_executor()
    HostExecution.handle(executor, "1", "not a map")

    rejection = find(collected_metrics(), "flexdoc_execute_rejections_total")
    assert rejection.labels["reason"] == "body-malformed"
  end

  test "an unreachable target is an upstream failure and the gauge still balances" do
    # Listen then close so the port is almost certainly refused rather than open.
    {:ok, listener} = :gen_tcp.listen(0, [:binary, active: false])
    {:ok, port} = :inet.port(listener)
    :ok = :gen_tcp.close(listener)
    origin = "http://127.0.0.1:#{port}"

    executor = collecting_executor([origin])
    result = HostExecution.handle(executor, "1", %{"request" => %{"url" => "#{origin}/pets", "method" => "GET"}})
    assert result.status == 502

    metrics = collected_metrics()
    failure = find(metrics, "flexdoc_execute_errors_total")
    assert failure.labels["reason"] == "upstream-unreachable"
    refute find(metrics, "flexdoc_execute_rejections_total")

    gauge =
      metrics
      |> Enum.filter(&(&1.name == "flexdoc_execute_in_flight"))
      |> Enum.map(& &1.value)
      |> Enum.sum()

    assert gauge == 0
  end

  test "a raising sink cannot fail an execution" do
    executor =
      HostExecution.new!(["https://api.example.test"],
        metric_sink: fn _metric -> raise "collector down" end
      )

    result =
      HostExecution.handle(executor, "1", %{
        "request" => %{"url" => "https://blocked.example.test/pets", "method" => "GET"}
      })

    assert result.status == 403
  end

  test "an executor without a sink behaves exactly as before" do
    executor = HostExecution.new!(["https://api.example.test"])
    assert HostExecution.handle(executor, nil, %{}).status == 403
  end

  test "the recorder aggregates by outcome and reason" do
    recorder = start_recorder()

    HostExecutionObservation.record(recorder, metric("flexdoc_execute_requests_total", :counter, 1))
    HostExecutionObservation.record(recorder, metric("flexdoc_execute_in_flight", :gauge, 1))
    HostExecutionObservation.record(recorder, metric("flexdoc_execute_in_flight", :gauge, 1))
    HostExecutionObservation.record(recorder, metric("flexdoc_execute_in_flight", :gauge, -1))

    HostExecutionObservation.record(
      recorder,
      metric("flexdoc_execute_completions_total", :counter, 1, %{"outcome" => "rejected"})
    )

    HostExecutionObservation.record(
      recorder,
      metric("flexdoc_execute_rejections_total", :counter, 1, %{"reason" => "destination-forbidden"})
    )

    HostExecutionObservation.record(
      recorder,
      metric("flexdoc_execute_errors_total", :counter, 1, %{"reason" => "upstream-timeout"})
    )

    HostExecutionObservation.record(
      recorder,
      metric("flexdoc_execute_unmarked_total", :counter, 1, %{"reason" => "marker-missing"})
    )

    HostExecutionObservation.record(
      recorder,
      metric("flexdoc_execute_errors_total", :counter, 1, %{"reason" => "not-a-reason"})
    )

    snapshot = settled_snapshot(recorder)
    assert snapshot["startedExecutions"] == 1
    assert snapshot["completedExecutions"] == 1
    assert snapshot["outcomes"]["rejected"] == 1
    assert snapshot["outcomes"]["success"] == 0
    assert snapshot["inFlight"] == 1
    assert snapshot["peakInFlight"] == 2
    assert snapshot["rejectionsByReason"]["destination-forbidden"] == 1
    assert map_size(snapshot["errorsByReason"]) == 1, "an unknown reason must not be tallied"
    assert snapshot["unmarkedRequests"] == 1
    refute snapshot["durations"], "no duration summary before the first completion duration"
  end

  test "the recorder reports exact percentiles" do
    recorder = start_recorder()

    Enum.each(1..100, fn index ->
      HostExecutionObservation.record(
        recorder,
        metric("flexdoc_execute_duration_seconds", :histogram, index / 1000, %{"outcome" => "success"})
      )
    end)

    durations = settled_snapshot(recorder)["durations"]
    refute durations["sampled"], "100 observations fit under the capacity"
    assert durations["sampleCount"] == 100
    assert_in_delta durations["minMs"], 1.0, 0.001
    assert_in_delta durations["p50Ms"], 50.0, 0.001
    assert_in_delta durations["p95Ms"], 95.0, 0.001
    assert_in_delta durations["p99Ms"], 99.0, 0.001
    assert_in_delta durations["maxMs"], 100.0, 0.001
  end

  test "the recorder bounds memory by sampling" do
    recorder = start_recorder(duration_sample_capacity: 16)

    Enum.each(1..500, fn _index ->
      HostExecutionObservation.record(recorder, metric("flexdoc_execute_duration_seconds", :histogram, 0.05))
    end)

    durations = settled_snapshot(recorder)["durations"]
    assert durations["sampleCount"] == 16, "capacity must bound retention"
    assert durations["sampled"], "sampling must be declared past capacity"
  end

  test "reset starts a new window" do
    recorder = start_recorder()
    HostExecutionObservation.record(recorder, metric("flexdoc_execute_requests_total", :counter, 1))
    :ok = HostExecutionObservation.reset(recorder)

    snapshot = settled_snapshot(recorder)
    assert snapshot["startedExecutions"] == 0
    refute snapshot["windowStart"]
  end

  test "the recorder sink is usable directly as an executor metric sink" do
    recorder = start_recorder()
    executor = HostExecution.new!(["https://api.example.test"], metric_sink: HostExecutionObservation.sink(recorder))
    HostExecution.handle(executor, nil, %{})

    assert settled_snapshot(recorder)["unmarkedRequests"] == 1
  end

  # The export is handed to operators and may be written to disk, so it must carry
  # no request content: only counts, timestamps and known category names.
  test "the report carries no request content" do
    recorder = start_recorder()
    HostExecutionObservation.record(recorder, metric("flexdoc_execute_requests_total", :counter, 1))

    HostExecutionObservation.record(
      recorder,
      metric("flexdoc_execute_rejections_total", :counter, 1, %{
        "reason" => "destination-forbidden",
        "target" => "https://secret.internal/pets?token=abc"
      })
    )

    HostExecutionObservation.record(recorder, metric("flexdoc_execute_duration_seconds", :histogram, 0.25))

    report = HostExecutionObservation.report(recorder)
    assert report["schema"] == "flexdoc.host-execution.observation/1"
    assert report["runtime"] == "elixir"
    assert report["gaps"] == ["browser-direct-transport-mix"]

    rendered = inspect(report, limit: :infinity)

    for leaked <- ["secret.internal", "token=abc", "/pets"] do
      refute String.contains?(rendered, leaked), "report leaked request content #{leaked}"
    end
  end
end

File: adapters/go/host_execution_observability_test.go
package flexdoc

import (
	"encoding/json"
	"net"
	"net/http"
	"net/http/httptest"
	"strings"
	"testing"
	"time"
)

// The reason vocabulary is a cross-runtime contract, not a Go detail: a collector
// written against the Node or Python host must read these labels unchanged.
func TestHostExecutionReasonsMatchTheSharedVocabulary(t *testing.T) {
	expected := []string{
		"marker-missing", "execution-disabled", "admission-saturated",
		"destination-forbidden", "redirect-forbidden", "body-malformed",
		"body-too-large", "unsupported-media-type", "request-invalid",
		"auth-unsupported", "upstream-timeout", "upstream-unreachable",
		"upstream-error",
	}
	reasons := HostExecutionReasons()
	if len(reasons) != len(expected) {
		t.Fatalf("expected %d reasons, got %d", len(expected), len(reasons))
	}
	for i, reason := range reasons {
		if string(reason) != expected[i] {
			t.Fatalf("reason %d: expected %q, got %q", i, expected[i], reason)
		}
	}
	if !IsHostExecutionReason("upstream-timeout") {
		t.Fatal("expected upstream-timeout to be a known reason")
	}
	if IsHostExecutionReason("slow") {
		t.Fatal("expected an unknown reason to be rejected")
	}
}

func collectMetrics(t *testing.T) (*HostExecution, *[]HostExecutionMetric) {
	t.Helper()
	collected := []HostExecutionMetric{}
	executor, err := NewHostExecution(
		[]string{"https://api.example.test"},
		WithMetricSink(func(metric HostExecutionMetric) {
			collected = append(collected, metric)
		}),
	)
	if err != nil {
		t.Fatalf("executor: %v", err)
	}
	return executor, &collected
}

func names(metrics []HostExecutionMetric) []string {
	out := make([]string, 0, len(metrics))
	for _, metric := range metrics {
		out = append(out, metric.Name)
	}
	return out
}

func findMetric(metrics []HostExecutionMetric, name string) (HostExecutionMetric, bool) {
	for _, metric := range metrics {
		if metric.Name == name {
			return metric, true
		}
	}
	return HostExecutionMetric{}, false
}

func TestUnmarkedRequestIsCountedOutsideTheLifecycle(t *testing.T) {
	executor, collected := collectMetrics(t)
	executor.Handle("", map[string]any{}, nil)

	if len(*collected) != 1 {
		t.Fatalf("expected exactly one metric, got %v", names(*collected))
	}
	metric := (*collected)[0]
	if metric.Name != "flexdoc_execute_unmarked_total" {
		t.Fatalf("expected the unmarked counter, got %q", metric.Name)
	}
	if metric.Labels["reason"] != "marker-missing" {
		t.Fatalf("expected marker-missing, got %q", metric.Labels["reason"])
	}
}

func TestPolicyRejectionCarriesItsReason(t *testing.T) {
	executor, collected := collectMetrics(t)
	result := executor.Handle("1", map[string]any{
		"request": map[string]any{"url": "https://blocked.example.test/pets", "method": "GET"},
	}, nil)

	if result.Status != http.StatusForbidden {
		t.Fatalf("expected 403, got %d", result.Status)
	}
	rejection, ok := findMetric(*collected, "flexdoc_execute_rejections_total")
	if !ok {
		t.Fatalf("expected a rejection counter, got %v", names(*collected))
	}
	if rejection.Labels["reason"] != "destination-forbidden" {
		t.Fatalf("expected destination-forbidden, got %q", rejection.Labels["reason"])
	}
	if rejection.Labels["statusCode"] != "403" || rejection.Labels["source"] != "route" {
		t.Fatalf("unexpected rejection labels: %v", rejection.Labels)
	}
	if _, errored := findMetric(*collected, "flexdoc_execute_errors_total"); errored {
		t.Fatal("a policy rejection must not also count as an upstream error")
	}
}

func TestMalformedEnvelopeIsSeparableFromAPolicyRejection(t *testing.T) {
	executor, collected := collectMetrics(t)
	executor.Handle("1", nil, nil)

	rejection, ok := findMetric(*collected, "flexdoc_execute_rejections_total")
	if !ok {
		t.Fatalf("expected a rejection counter, got %v", names(*collected))
	}
	if rejection.Labels["reason"] != "body-malformed" {
		t.Fatalf("expected body-malformed, got %q", rejection.Labels["reason"])
	}
}

func TestUnreachableTargetIsAnUpstreamFailureNotARejection(t *testing.T) {
	listener, err := net.Listen("tcp", "127.0.0.1:0")
	if err != nil {
		t.Fatalf("listener: %v", err)
	}
	address := listener.Addr().String()
	// Close immediately so the port is almost certainly refused rather than open.
	_ = listener.Close()

	collected := []HostExecutionMetric{}
	executor, err := NewHostExecution(
		[]string{"http://" + address},
		WithMetricSink(func(metric HostExecutionMetric) { collected = append(collected, metric) }),
	)
	if err != nil {
		t.Fatalf("executor: %v", err)
	}
	result := executor.Handle("1", map[string]any{
		"request": map[string]any{"url": "http://" + address + "/pets", "method": "GET"},
	}, nil)

	if result.Status != http.StatusBadGateway {
		t.Fatalf("expected 502, got %d", result.Status)
	}
	failure, ok := findMetric(collected, "flexdoc_execute_errors_total")
	if !ok {
		t.Fatalf("expected an error counter, got %v", names(collected))
	}
	if failure.Labels["reason"] != "upstream-unreachable" {
		t.Fatalf("expected upstream-unreachable, got %q", failure.Labels["reason"])
	}
	if _, rejected := findMetric(collected, "flexdoc_execute_rejections_total"); rejected {
		t.Fatal("an upstream failure must not be counted as a policy rejection")
	}
}

func TestSuccessfulExecutionEmitsTheLifecyclePair(t *testing.T) {
	target := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, _ *http.Request) {
		w.Header().Set("Content-Type", "application/json")
		_, _ = w.Write([]byte(`{"ok":true}`))
	}))
	defer target.Close()

	collected := []HostExecutionMetric{}
	executor, err := NewHostExecution(
		[]string{target.URL},
		WithMetricSink(func(metric HostExecutionMetric) { collected = append(collected, metric) }),
	)
	if err != nil {
		t.Fatalf("executor: %v", err)
	}
	result := executor.Handle("1", map[string]any{
		"request": map[string]any{"url": target.URL + "/pets", "method": "GET"},
	}, nil)
	if result.Status != http.StatusOK {
		t.Fatalf("expected 200, got %d: %v", result.Status, result.Body)
	}

	if _, ok := findMetric(collected, "flexdoc_execute_requests_total"); !ok {
		t.Fatalf("expected a requests counter, got %v", names(collected))
	}
	completion, ok := findMetric(collected, "flexdoc_execute_completions_total")
	if !ok || completion.Labels["outcome"] != "success" {
		t.Fatalf("expected a successful completion, got %v", names(collected))
	}
	if _, ok := findMetric(collected, "flexdoc_execute_duration_seconds"); !ok {
		t.Fatal("expected a duration observation")
	}

	// The gauge must return to zero, or in-flight drifts upward forever.
	var gauge float64
	for _, metric := range collected {
		if metric.Name == "flexdoc_execute_in_flight" {
			gauge += metric.Value
		}
	}
	if gauge != 0 {
		t.Fatalf("expected the in-flight gauge to balance, got %v", gauge)
	}
}

func TestAFailingSinkCannotFailAnExecution(t *testing.T) {
	executor, err := NewHostExecution(
		[]string{"https://api.example.test"},
		WithMetricSink(func(HostExecutionMetric) { panic("collector down") }),
	)
	if err != nil {
		t.Fatalf("executor: %v", err)
	}
	result := executor.Handle("1", map[string]any{
		"request": map[string]any{"url": "https://blocked.example.test/pets", "method": "GET"},
	}, nil)
	if result.Status != http.StatusForbidden {
		t.Fatalf("expected the execution outcome to survive a panicking sink, got %d", result.Status)
	}
}

func TestNoMetricsWithoutASink(t *testing.T) {
	executor, err := NewHostExecution([]string{"https://api.example.test"})
	if err != nil {
		t.Fatalf("executor: %v", err)
	}
	// Nothing to assert beyond not panicking: an executor without a sink must
	// behave exactly as it did before evidence existed.
	if result := executor.Handle("", map[string]any{}, nil); result.Status != http.StatusForbidden {
		t.Fatalf("expected 403, got %d", result.Status)
	}
}

func recorderWithClock() (*HostExecutionObservation, *time.Time) {
	instant := time.Date(2026, 9, 21, 8, 0, 0, 0, time.UTC)
	observation := NewHostExecutionObservation()
	observation.now = func() time.Time { return instant }
	return observation, &instant
}

func TestObservationAggregatesByOutcomeAndReason(t *testing.T) {
	observation, _ := recorderWithClock()
	observation.Record(HostExecutionMetric{Name: "flexdoc_execute_requests_total", Kind: "counter", Value: 1})
	observation.Record(HostExecutionMetric{Name: "flexdoc_execute_in_flight", Kind: "gauge", Value: 1})
	observation.Record(HostExecutionMetric{Name: "flexdoc_execute_in_flight", Kind: "gauge", Value: 1})
	observation.Record(HostExecutionMetric{Name: "flexdoc_execute_in_flight", Kind: "gauge", Value: -1})
	observation.Record(HostExecutionMetric{Name: "flexdoc_execute_completions_total", Kind: "counter", Value: 1, Labels: map[string]string{"outcome": "rejected"}})
	observation.Record(HostExecutionMetric{Name: "flexdoc_execute_rejections_total", Kind: "counter", Value: 1, Labels: map[string]string{"reason": "destination-forbidden"}})
	observation.Record(HostExecutionMetric{Name: "flexdoc_execute_errors_total", Kind: "counter", Value: 1, Labels: map[string]string{"reason": "upstream-timeout"}})
	observation.Record(HostExecutionMetric{Name: "flexdoc_execute_unmarked_total", Kind: "counter", Value: 1, Labels: map[string]string{"reason": "marker-missing"}})
	observation.Record(HostExecutionMetric{Name: "flexdoc_execute_errors_total", Kind: "counter", Value: 1, Labels: map[string]string{"reason": "not-a-reason"}})

	snapshot := observation.Snapshot()
	if snapshot.StartedExecutions != 1 || snapshot.CompletedExecutions != 1 {
		t.Fatalf("unexpected counts: %+v", snapshot)
	}
	if snapshot.Outcomes["rejected"] != 1 || snapshot.Outcomes["success"] != 0 {
		t.Fatalf("unexpected outcomes: %v", snapshot.Outcomes)
	}
	if snapshot.InFlight != 1 || snapshot.PeakInFlight != 2 {
		t.Fatalf("expected peak concurrency 2 with 1 still running, got %d/%d", snapshot.InFlight, snapshot.PeakInFlight)
	}
	if snapshot.RejectionsByReason["destination-forbidden"] != 1 {
		t.Fatalf("unexpected rejections: %v", snapshot.RejectionsByReason)
	}
	if snapshot.ErrorsByReason["upstream-timeout"] != 1 || len(snapshot.ErrorsByReason) != 1 {
		t.Fatalf("an unknown reason must not be tallied: %v", snapshot.ErrorsByReason)
	}
	if snapshot.UnmarkedRequests != 1 {
		t.Fatalf("expected the unmarked count to stay separate, got %d", snapshot.UnmarkedRequests)
	}
	if snapshot.Durations != nil {
		t.Fatal("expected no duration summary before the first completion duration")
	}
}

func TestObservationReportsExactPercentiles(t *testing.T) {
	observation, _ := recorderWithClock()
	for i := 1; i <= 100; i++ {
		observation.Record(HostExecutionMetric{
			Name:   "flexdoc_execute_duration_seconds",
			Kind:   "histogram",
			Value:  float64(i) / 1000,
			Labels: map[string]string{"outcome": "success"},
		})
	}
	durations := observation.Snapshot().Durations
	if durations == nil {
		t.Fatal("expected a duration summary")
	}
	if durations.Sampled {
		t.Fatal("100 observations fit under the capacity and must not be reported as sampled")
	}
	if durations.MinMs != 1 || durations.MaxMs != 100 || durations.P50Ms != 50 || durations.P95Ms != 95 || durations.P99Ms != 99 {
		t.Fatalf("unexpected percentiles: %+v", durations)
	}
}

func TestObservationBoundsMemoryBySampling(t *testing.T) {
	observation := NewHostExecutionObservationWithCapacity(16)
	observation.now = func() time.Time { return time.Unix(0, 0).UTC() }
	for i := 0; i < 500; i++ {
		observation.Record(HostExecutionMetric{Name: "flexdoc_execute_duration_seconds", Kind: "histogram", Value: 0.05})
	}
	durations := observation.Snapshot().Durations
	if durations.SampleCount != 16 {
		t.Fatalf("expected the capacity to bound retention, got %d", durations.SampleCount)
	}
	if !durations.Sampled {
		t.Fatal("expected sampling to be declared once observations exceed capacity")
	}
}

func TestObservationResetStartsANewWindow(t *testing.T) {
	observation, _ := recorderWithClock()
	observation.Record(HostExecutionMetric{Name: "flexdoc_execute_requests_total", Kind: "counter", Value: 1})
	observation.Reset()

	snapshot := observation.Snapshot()
	if snapshot.StartedExecutions != 0 || snapshot.WindowStart != nil {
		t.Fatalf("expected a cleared window, got %+v", snapshot)
	}
}

// The export is handed to operators and may be written to disk, so it must carry
// no request content: only counts, timestamps and known category names.
func TestReportCarriesNoRequestContent(t *testing.T) {
	observation, _ := recorderWithClock()
	observation.Record(HostExecutionMetric{Name: "flexdoc_execute_requests_total", Kind: "counter", Value: 1})
	observation.Record(HostExecutionMetric{
		Name:   "flexdoc_execute_rejections_total",
		Kind:   "counter",
		Value:  1,
		Labels: map[string]string{"reason": "destination-forbidden", "target": "https://secret.internal/pets?token=abc"},
	})
	observation.Record(HostExecutionMetric{Name: "flexdoc_execute_duration_seconds", Kind: "histogram", Value: 0.25})

	report := NewHostExecutionObservationReport(observation)
	if report.Schema != "flexdoc.host-execution.observation/1" {
		t.Fatalf("unexpected schema: %q", report.Schema)
	}
	if report.Runtime != "go" {
		t.Fatalf("expected the runtime to identify itself, got %q", report.Runtime)
	}
	if len(report.Gaps) != 1 || report.Gaps[0] != "browser-direct-transport-mix" {
		t.Fatalf("expected the declared transport-mix gap, got %v", report.Gaps)
	}

	encoded, err := json.Marshal(report)
	if err != nil {
		t.Fatalf("report must be serializable: %v", err)
	}
	for _, leaked := range []string{"secret.internal", "token=abc", "/pets"} {
		if strings.Contains(string(encoded), leaked) {
			t.Fatalf("report leaked request content %q: %s", leaked, encoded)
		}
	}
}
